libs/contrib/Data/Fun/Extra.idr

```diff
@@ -0,0 +1,54 @@
+module Data.Fun.Extra
+
+import Data.Fun
+import Data.HVect
+
+||| Apply an n-ary function to an n-ary tuple of inputs
+public export
+apply : {ts : Vect n Type} -> {cod : Type} -> Fun ts cod -> HVect ts -> cod
+apply f [] = f
+apply f (t::ts) = apply (f t) ts
+
+||| Apply an n-ary function to an n-ary tuple of inputs
+public export
+applyPartially : {ts : Vect n Type} -> {ss : Vect m Type} -> {rs : Vect k Type} 
+               -> {auto 0 fordShape : rs = ts ++ ss}
+               -> {cod : Type} -> Fun rs cod -> HVect ts -> Fun ss cod
+applyPartially {fordShape = Refl} f [] = f
+applyPartially {fordShape = Refl} f (x::xs) = applyPartially (f x) xs
+
+
+{- -------- (slightly) dependent versions of the above ---------------
+   As usual, type dependencies make everything complicated          -}
+
+
+||| Build an n-ary dependent function type from a Vect of Types and a dependent result type
+||| NB: the types in the tuple do not depend on each other.
+public export
+GenFun : (ts : Vect n Type) -> (cod : Fun ts Type) -> Type
```

Comment:
This is `Data.Rel`'s `All`. I need to think harder about the stuff that
comes after that to understand what it corresponds to.

Reply:
I see, thanks! I would really like to have a fully dependently typed `GenFun`, but I've never managed to work out how to best represent telescopes (rather than `Vect n Type`).

I've renamed it, it seems weird to `import Data.Rel`, but I guess that's OK.